<h1 align=center><font size=6>Predicción del precio de un vehículo🚗</font></h1>

### Crear el dataframe

In [147]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.linear_model import LinearRegression

In [148]:
# Crear el dataframe
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"
cars = pd.read_csv(url, header=None)  

In [149]:
# Crear lista con los nombres de las columnas y reemplazar en el dataframe
headers = ["symboling","normalized-losses","make","fuel-type","aspiration","num-of-doors","body-style",
"drive-wheels","engine-location","wheel-base","length","width","height","curb-weight", "engine-type",
"num-of-cylinders","engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
"peak-rpm","city-mpg","highway-mpg","price"]
cars.columns = headers

In [125]:
# Imprimir las primeras cinco filas
cars.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

### Identificar los valores faltantes

Algunas columnas tienen datos sin valor que aparecen como "?", se debe reemplazar por NaN el cual es el marcador por defecto de Python para valores faltantes por razones de conveniencia y velocidad de computo. 

In [126]:
cars.replace("?", np.nan, inplace=True)
cars.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [127]:
# Crear un dataframe de booleanos
missing_data = cars.isnull()
missing_data.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

In [128]:
# contar datos faltantes por columna
for column in missing_data:
    print(column)
    print(missing_data[column].value_counts()) 

symboling
False    205
Name: symboling, dtype: int64
normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64
make
False    205
Name: make, dtype: int64
fuel-type
False    205
Name: fuel-type, dtype: int64
aspiration
False    205
Name: aspiration, dtype: int64
num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64
body-style
False    205
Name: body-style, dtype: int64
drive-wheels
False    205
Name: drive-wheels, dtype: int64
engine-location
False    205
Name: engine-location, dtype: int64
wheel-base
False    205
Name: wheel-base, dtype: int64
length
False    205
Name: length, dtype: int64
width
False    205
Name: width, dtype: int64
height
False    205
Name: height, dtype: int64
curb-weight
False    205
Name: curb-weight, dtype: int64
engine-type
False    205
Name: engine-type, dtype: int64
num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64
engine-size
False    205
Name: engine-size, dtype: int64
fuel-system
False    205
Name: 

De acuerdo a lo anterior, tenemos entonces que todas las columnas tienen 205 filas, con datos faltantes en las siguientes:</p> 

- "normalized-losses": 41 datos faltantes
- "num-of-doors": 2 datos faltantes
- "bore": 4 datos faltantes
- "stroke" : 4 datos faltantes
- "horsepower": 2 datos faltantes
- "peak-rpm": 2 datos faltantes
- "price": 4 datos faltantes


### Reemplazar o eliminar datos faltantes 
Primero podemos obtener una descripción de los datos y su tipo para ver su comportamiento.

In [129]:
cars.describe(include="all")

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               164     205       205        205   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           203        205          205             205  205.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   201     201        205.000000        203   
unique          NaN            8    38      36               NaN         59   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        203  205.000000  205.000000   201  
unique        23         NaN         NaN   186  
top         5500         NaN         NaN  8921  
freq          37         NaN         NaN     2  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

In [130]:
cars.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Podemos entonces decidir lo siquiente:</p> 

**Reemplazar con la media:**
- "normalized-losses"
- "bore"
- "stroke"
- "horsepower"
- "peak-rpm"
    
 </p> 
    
**Reemplazar con la frecuencia:** 
- "num-of-doors": se puede reemplazar con "four" ya que es el dato que más se repite
    </p> 
    
**Eliminar datos:**  
- "price": se eliminan las filas con datos faltantes ya que es la variable que queremos predecir. </p> 

**Nota:** Es importante al momento de calcular la media asignarle tipo "float", ya que tienen asigando erróneamente sus tipos.

In [143]:
# eliminar filas de la columna price
cars.dropna(subset=["price"], inplace=True)

# reemplazar con la media, asignar tipo "float" para que lo pueda calcular ya que tiene asignado erroneamente su tipo
avr_norm_loss = cars["normalized-losses"].astype("float64").mean()
cars["normalized-losses"].replace(np.nan, avr_norm_loss, inplace=True)

avr_bore = cars["bore"].astype("float64").mean()
cars["bore"].replace(np.nan, avr_bore, inplace=True)

avr_stroke = cars["stroke"].astype("float64").mean()
cars["stroke"].replace(np.nan, avr_stroke, inplace=True)

avr_horsepower = cars["horsepower"].astype("float64").mean()
cars["horsepower"].replace(np.nan, avr_horsepower, inplace=True)

avr_rpm = cars["peak-rpm"].astype("float64").mean()
cars["peak-rpm"].replace(np.nan, avr_rpm, inplace=True)

# reemplazar con la frecuencia
freq_num_doors = df['num-of-doors'].value_counts().idxmax()
cars["num-of-doors"].replace(np.nan, freq_num_doors, inplace=True)


###  Corregir el formato de datos:
Como se vió anteiormente, algunas columnas tienen un tipo de dato erróneo, así que se debe corregir entonces:
- "normalized-losses": object -> int64
- "bore": object -> float64
- "stroke": object -> float64
- "horsepower": object -> int64
- "peak-rpm": object -> int64
- "price": object -> int64


In [144]:
cars["normalized-losses"] = cars["normalized-losses"].astype("int64")
cars["bore"] = cars["bore"].astype("float64")
cars["stroke"] = cars["stroke"].astype("float64")
cars["horsepower"] = cars["horsepower"].astype("int64")
cars["peak-rpm"] = cars["peak-rpm"].astype("int64")
cars["price"] = cars["price"].astype("int64")

In [145]:
cars.describe(include="all")

symboling  normalized-losses    make fuel-type aspiration  \
count   201.000000          201.00000     201       201        201   
unique         NaN                NaN      22         2          2   
top            NaN                NaN  toyota       gas        std   
freq           NaN                NaN      32       181        165   
mean      0.840796          122.00000     NaN       NaN        NaN   
std       1.254802           31.99625     NaN       NaN        NaN   
min      -2.000000           65.00000     NaN       NaN        NaN   
25%       0.000000          101.00000     NaN       NaN        NaN   
50%       1.000000          122.00000     NaN       NaN        NaN   
75%       2.000000          137.00000     NaN       NaN        NaN   
max       3.000000          256.00000     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           201        201          201             201  201.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            115         94          118             198         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.797015  ...   
std             NaN        NaN          NaN             NaN    6.066366  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system        bore      stroke compression-ratio  \
count    201.000000          201  201.000000  201.000000        201.000000   
unique          NaN            8         NaN         NaN               NaN   
top             NaN         mpfi         NaN         NaN               NaN   
freq            NaN           92         NaN         NaN               NaN   
mean     126.875622          NaN    3.330711    3.256904         10.164279   
std       41.546834          NaN    0.268072    0.316048          4.004965   
min       61.000000          NaN    2.540000    2.070000          7.000000   
25%       98.000000          NaN    3.150000    3.110000          8.600000   
50%      120.000000          NaN    3.310000    3.290000          9.000000   
75%      141.000000          NaN    3.580000    3.410000          9.400000   
max      326.000000          NaN    3.940000    4.170000         23.000000   

        horsepower     peak-rpm    city-mpg  highway-mpg         price  
count   201.000000   201.000000  201.000000   201.000000    201.000000  
unique         NaN          NaN         NaN          NaN           NaN  
top            NaN          NaN         NaN          NaN           NaN  
freq           NaN          NaN         NaN          NaN           NaN  
mean    103.393035  5117.582090   25.179104    30.686567  13207.129353  
std      37.365623   478.113182    6.423220     6.815150   7947.066342  
min      48.000000  4150.000000   13.000000    16.000000   5118.000000  
25%      70.000000  4800.000000   19.000000    25.000000   7775.000000  
50%      95.000000  5117.000000   24.000000    30.000000  10295.000000  
75%     116.000000  5500.000000   30.000000    34.000000  16500.000000  
max     262.000000  6600.000000   49.000000    54.000000  45400.000000  

[11 rows x 26 columns]

symboling  normalized-losses         make fuel-type aspiration  \
0           3                122  alfa-romero       gas        std   
1           3                122  alfa-romero       gas        std   
2           1                122  alfa-romero       gas        std   
3           2                164         audi       gas        std   
4           2                164         audi       gas        std   
5           2                122         audi       gas        std   
6           1                158         audi       gas        std   
7           1                122         audi       gas        std   
8           1                158         audi       gas      turbo   
10          2                192          bmw       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0           two  convertible          rwd           front        88.6  ...   
1           two  convertible          rwd           front        88.6  ...   
2           two    hatchback          rwd           front        94.5  ...   
3          four        sedan          fwd           front        99.8  ...   
4          four        sedan          4wd           front        99.4  ...   
5           two        sedan          fwd           front        99.8  ...   
6          four        sedan          fwd           front       105.8  ...   
7          four        wagon          fwd           front       105.8  ...   
8          four        sedan          fwd           front       105.8  ...   
10          two        sedan          rwd           front       101.2  ...   

    engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0           130         mpfi  3.47    2.68               9.0        111   
1           130         mpfi  3.47    2.68               9.0        111   
2           152         mpfi  2.68    3.47               9.0        154   
3           109         mpfi  3.19    3.40              10.0        102   
4           136         mpfi  3.19    3.40               8.0        115   
5           136         mpfi  3.19    3.40               8.5        110   
6           136         mpfi  3.19    3.40               8.5        110   
7           136         mpfi  3.19    3.40               8.5        110   
8           131         mpfi  3.13    3.40               8.3        140   
10          108         mpfi  3.50    2.80               8.8        101   

    peak-rpm city-mpg  highway-mpg  price  
0       5000       21           27  13495  
1       5000       21           27  16500  
2       5000       19           26  16500  
3       5500       24           30  13950  
4       5500       18           22  17450  
5       5500       19           25  15250  
6       5500       19           25  17710  
7       5500       19           25  18920  
8       5500       17           20  23875  
10      5800       23           29  16430  

[10 rows x 26 columns]